Implement `SimpleReplace`

cqc-alec opened the pull request:
This PR includes and implements #85 .

Since we don't yet have DFG-rooted HUGRs we use a workaround where the replacement graph is specified as a HUGR with a unique DFG node. Once #112 is merged we can remove this workaround.

acl-cqc commented on src/hugr.rs:
```diff
@@ -73,6 +77,127 @@ impl Hugr {
+            for n_node_succ in r.n.output_neighbours(n_node) {
+                if r.n.get_optype(n_node_succ).tag() != OpTag::Output {
+                    let new_node_succ_index = index_map.get(&n_node_succ.index).unwrap();
+                    for connection in r.n.graph.get_connections(*n_index, n_node_succ.index) {
```
Hmmm, there is a lot of `.index` going on here - it's strange that PortGraph `port_offset` returns an `Option<PortOffset>` but `link_nodes` takes `usize`s (that it converts to PortOffsets internally).

I wonder if you can do this more easily with `HugrView::node_outputs`...ah, but we can't mutate the Hugr while we hold a view, so we can't use that.

Bah. Ok, fine. Might be worth a PR to update PortGraph.link_nodes to use PortOffset, but I'm not gonna hold this up for that.

aborgna-q replied:
> it's strange that PortGraph port_offset returns an Option<PortOffset> but link_nodes takes usizes

Yes, I'm changing that soon™️

acl-cqc commented on src/hugr.rs:
```diff
@@ -73,6 +77,127 @@ impl Hugr {
+                .ok();
+            index_map.insert(*n_index, new_node_index);
+        }
+        // 3.2. Add edges between all newly added nodes matching those in n_dfg_node.
```
IIUC the spec was written assuming implicit copies so I'm guessing this may change when that's ready.

acl-cqc commented on src/replacement/simple_replace.rs:
```diff
@@ -0,0 +1,228 @@
+pub struct SimpleReplacement {
+    /// The common DFG parent of all nodes to be replaced.
+    pub p: Node,
+    /// The set of nodes to remove (a convex set of leaf children of `p`).
```
How about calling this `remove` (with comment "called S in the spec" if you like!)

cqc-alec replied:
I propose renaming:

* `p` --> `region`
* `s` --> `removal`
* `n` --> `replacement`

acl-cqc replied:
I love `removal` and agree with `replacement`. Less sure about `region` - maybe just `parent`?

cqc-alec replied:
Changed to `parent`.

acl-cqc commented on src/replacement/simple_replace.rs:
```diff
@@ -0,0 +1,228 @@
+
+    const QB: SimpleType = SimpleType::Linear(LinearType::Qubit);
+
+    /// Creates a hugr with a DFG.
```
When we have rooted-at-any-op Hugrs, we might as well skip much of this as well, right? (Just to simplify the test)

acl-cqc commented on src/replacement/simple_replace.rs:
```diff
@@ -0,0 +1,228 @@
+            let [qb0, qb1] = func_builder.input_wires_arr();
+
+            let mut inner_builder =
+                func_builder.dfg_builder(vec![(QB, qb0), (QB, qb1)], type_row![QB, QB])?;
```
and ASCII art again

aborgna-q commented on src/hugr.rs:
```diff
@@ -73,6 +77,127 @@ impl Hugr {
+                .link_ports(new_out_portindex, s_exit_portindex)
+                .ok();
+        }
+        // 3.5. Remove all nodes in r.s and edges between them.
```
Can we do this before inserting the new nodes, to allow reusing memory?

cqc-alec replied:
No, because we then panic at the line
```
let rem_inp_predecessor_port_index = self.graph.port_link(rem_inp_port_index).unwrap();
```
where we need to locate the predecessor node.

aborgna-q reviewed (changes requested):
I strongly suggest using more meaningful names in the implementation. Having to go to an internal definition in another module to parse what `r.s` adds a lot of friction.

cqc-alec commented:
Addressed comments in the PR (except the ones about code location, which I'll return to). Moved to draft since it implements the old version of `SimpleReplace` (which didn't handle empty wires in the replacement), so needs updating to match the new version...

acl-cqc commented on src/hugr.rs:
This is in progress in #119

cqc-alec commented:
I think this now handles empty wires correctly. Note the updated spec.

cqc-alec commented:
Some fixes needed following merge from latest `main`; update to follow...

aborgna-q commented on src/hugr.rs:
```diff
@@ -70,6 +75,193 @@ pub struct Wire(Node, usize);
+        let mut index_map: HashMap<NodeIndex, NodeIndex> = HashMap::new();
+        let replacement_nodes = r
+            .replacement
+            .hierarchy
+            .children(r.replacement.root().index)
+            .map_into::<Node>()
```
```suggestion
            .children(r.replacement.root())
```

aborgna-q commented on src/hugr.rs:
```diff
@@ -70,6 +75,193 @@ pub struct Wire(Node, usize);
+            }
+        }
+        let self_input_node_index = self.hierarchy.first(r.parent.index).unwrap();
+        let replacement_output_node = replacement_nodes[replacement_sz - 1];
```
Nitpicky, but it's slightly clearer
```suggestion
        let replacement_output_node = *replacement_nodes.last().unwrap();
```

aborgna-q commented on src/hugr.rs:
Is the idea to move `apply_simple_replacement` to `simple_replace.rs` once #119 is done?
The code could be moved over there now, so it's less noisy to change that afterwards.

cqc-alec replied:
Yeah I was going to wait until that was done, since there isn't really a framework for it yet.

aborgna-q reviewed (commented):
`Result::ok` just casts the value to an `Option`, discarding any error information. In most cases calling that before `unwrap` will give less debug info.

cqc-alec commented on specification/hugr.md:
```diff
@@ -1090,16 +1092,20 @@ The method takes as input:
   - a map $\nu\_\textrm{inp}: \textrm{inp}\_H(T \setminus \\{\texttt{Input}\\}) \to \textrm{inp}\_{\Gamma}(S)$;
   - a map $\nu_\textrm{out}: \textrm{out}_{\Gamma}(S) \to \textrm{out}_H(T \setminus \\{\texttt{Output}\\})$.
```
I think they are correct? T is a set in H, and S is a set in Gamma.

aborgna-q replied:
Just above it says:

> Given a set $S$ of nodes in a hugr $H$, let:

cqc-alec replied:
I see the confusion. But that is a different H (scoped to the definition). I will change the definition to use a different letter!

cqc-alec commented:
> `Result::ok` just casts the value to an `Option`, discarding any error information. In most cases calling that before `unwrap` will give less debug info.

Thanks, yes. It seems I can't change all instances of `.ok().unwrap()` though because of errors like:

```
error[E0277]: `NodePortOffsets` doesn't implement `Debug`
    --> src/replacement/simple_replace.rs:230:14
     |
230  |             .unwrap();
     |              ^^^^^^ `NodePortOffsets` cannot be formatted using `{:?}` because it doesn't implement `Debug`
     |
     = help: the trait `Debug` is not implemented for `NodePortOffsets`
     = help: the trait `Debug` is implemented for `ExactlyOneError<I>`
     = note: required for `itertools::adaptors::map::MapSpecialCase<NodePortOffsets, itertools::adaptors::map::MapSpecialCaseFnInto<Port>>` to implement `Debug`
     = note: 1 redundant requirement hidden
     = note: required for `ExactlyOneError<MapSpecialCase<NodePortOffsets, MapSpecialCaseFnInto<Port>>>` to implement `Debug`
     = note: the full type name has been written to '/home/alec/r/hugr/target/debug/deps/hugr-653fda1ba5647c44.long-type-8370086427390340701.txt'
note: required by a bound in `Result::<T, E>::unwrap`
    --> /home/alec/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1085:12
     |
1085 |         E: fmt::Debug,
     |            ^^^^^^^^^^ required by this bound in `Result::<T, E>::unwrap`
```

Will change what I can.
